benchmarks/ProtoBenchCaliper.java

```diff
@@ -0,0 +1,226 @@
+  @BeforeExperiment
+  void setUp() throws IOException {
+    inputData = ProtoBench.readAllBytes(dataFile);
+    benchmarkDataset = BenchmarkDataset.parseFrom(inputData);
+    defaultMessage = benchmarkMessageType.GetDefaultInstance();
+    extensions = benchmarkMessageType.registerExtension();
+    inputDataList = new ArrayList<byte[]>();
+    inputStreamList = new ArrayList<ByteArrayInputStream>();
+    inputStringList = new ArrayList<ByteString>();
+    sampleMessageList = new ArrayList<Message>();
+    
+    for (int i = benchmarkDataset.getPayloadCount() - 1; i >= 0; i--) {
+      byte[] singleInputData = benchmarkDataset.getPayload(i).toByteArray();
+      inputDataList.add(benchmarkDataset.getPayload(i).toByteArray());
+      inputStreamList.add(new ByteArrayInputStream(
+          benchmarkDataset.getPayload(i).toByteArray()));
+      inputStringList.add(benchmarkDataset.getPayload(i));
+      sampleMessageList.add(
+          defaultMessage.newBuilderForType().mergeFrom(singleInputData, extensions).build());
+    }
+    devNull = null;
+    reuseDevNull = null;
+    
+    try {
+      devNull = new FileOutputStream("/dev/null");
+      reuseDevNull = CodedOutputStream.newInstance(devNull);
+    } catch (FileNotFoundException e) {
+      // ignore: this is probably Windows, where /dev/null does not exist
+    }
+    
+    counter = 0;
+  }
+  
+  
+  @Benchmark
+  void SerializeToByteString(int reps) throws IOException {
+    for (int i = 0; i < reps; i++) {
+      sampleMessageList.get(counter % sampleMessageList.size()).toByteString();
+      counter++;
+    }
+  }
+  
+  @Benchmark
+  void SerialzeToByteArray(int reps) throws IOException {
+    for (int i = 0; i < reps; i++) {
+      sampleMessageList.get(counter % sampleMessageList.size()).toByteArray();
+      counter++;
+    }
+  }
+  
+  @Benchmark
+  void SerialzeToMemoryStream(int reps) throws IOException {
+    for (int i = 0; i < reps; i++) {
+      ByteArrayOutputStream output = new ByteArrayOutputStream();
+      sampleMessageList.get(counter % sampleMessageList.size()).writeTo(output);
+      counter++;
+    }
+  }
+  
+  @Benchmark
+  void DeserializeFromByteString(int reps) throws IOException {
+    for (int i = 0; i < reps; i++) {
+      defaultMessage
+        .newBuilderForType()
+        .mergeFrom(inputStringList.get(counter % inputStringList.size()), extensions)
+        .build();
+      counter++;
+    }
+  }
+  
+  @Benchmark
+  void DeserializeFromByteArray(int reps) throws IOException {
+    for (int i = 0; i < reps; i++) {
+      defaultMessage
+        .newBuilderForType()
+        .mergeFrom(inputDataList.get(counter % inputDataList.size()), extensions)
+        .build();
+      counter++;
+    }
+  }
+  
+  @Benchmark
+  void DeserializeFromMemoryStream(int reps) throws IOException {
+    for (int i = 0; i < reps; i++) {
+      defaultMessage
+        .newBuilderForType()
+        .mergeFrom(inputStreamList.get(counter % inputStreamList.size()), extensions)
```

Comment:
This seems wrong. The InputStream will persist its state across multiple parsing passes and all except the first will do nothing because the InputStream is already empty.

Reply:
I run reset after serialization, From the result 'reset' function should be correct